### PR TITLE
Add conditional visibility rules to settings list

### PR DESCRIPTION
**Pull request by LinkofOrigin**
# CURRENT STATE (WIP)

Still actively working on this. Creating this PR after getting just one scenario working, plenty needs addressed still but I think it works well enough as a proof-of-concept.

Here's some things I can think of so far that need addressed:
- [ ] Update remaining option types to support "conditional_visibility" (currently only added to Scale)
- [ ] Figure out how to deal with multi-value settings that should be disabled/enabled by one setting, but are also dependencies for displaying other settings. Example: If you set ShuffleGanonBosskey to "triforces" (new value), you can see Triforce*PerWorld settings. Then if you enable TriforceHunt, it will hide all the scalar settings tied to the Ganon Boss Key (Dungeon Rewards, Hearts, etc.). Should that still be disabled? Should we allow it to be enabled since the Triforce*PerWorld settings are displayed? Speaking of...
- [ ] Setting combination priority: Not really in the scope of this change, but if there are multiple setting combinations currently enabled, then something needs to decide what to do. The "simple" option is to just not decide and display an error to the user. Could also give priority to certain combinations and just ignore others. Example: If TriforceHunt is enabled AND ShuffleGanonBosskey is set to "triforces", then we could choose to respect TriforceHunt and ignore the fact that ShuffleGanonBosskey also happens to rely on triforce pieces.
- [ ] Needs testing: What if a setting relies on one or more other settings of which none have any "disable" or "conditional_visibility" json? Does this still get processed properly?
- [ ] Add conditional "hide" logic as well. Currently it can only conditionally show a setting if at least one condition is met. There is nothing that
- [ ] Consider "prioritization" logic to the condition list. Currently all conditions are treated equal, and if one passes then the setting is shown. Might be pertinent to add logic that allows a condition to pass and set the current state without processing other conditions. Many ways to approach that...

# Conditional Visibility Feature

_GitHub Issue: https://github.com/OoTRandomizer/OoT-Randomizer/issues/1827#issue-1496446474_

As it stands today, relationships between settings is limited to "disable only" per value. That is, if a setting is set to particular value, that value can "disable" other settings from being altered in the GUI.

This change aims to both reduce rigidity in the existing "disable" logic as well as open up the possibility of more complex logic. For example, a setting option may be able to _enable_ another setting, or a certain setting can be reliant on a combination of other setting values to be enabled (or disabled).

**Example**: TriforceHunt is a Checkbox (true/false) and currently disables ShuffleGanonBosskey when set to "true". When "true", TriforceHunt also enables two other settings: TriforceCountPerWorld and TriforceGoalPerWorld. There is currently no way to enable those secondary "Triforce*PerWorld" settings without TriforceHunt being enabled because the "disable" logic says so.

Let's say I want to make a new feature where Ganon's Boss Key can be acquired directly off of triforce pieces being collected WITHOUT enabling TriforceHunt. (Yes, I know this is technically what happens today, but bear with me a moment.) This way acquiring the last triforce piece only gives the boss key and does NOT trigger the outro cutscene, requiring the player to still defeat Ganon to "win" the seed (ie. bridge requirements and "kill ganon" requirements must also be met).

This new conditional logic allows this to happen in the GUI. Simply do the following:
- Add "conditional_visibility" json to the "Triforce*PerWorld" entries in the SettingsList. Set the logic such that these settings will display when TriforceHunt is enabled OR when both TriforceHunt is disabled AND ShuffleGanonBosskey is set to a new "triforces" option.
- Remove the "disable" json from the current TriforceHunt entry in the SettingsList.

Now someone just needs to update the rom patching logic to support this setting combination :)

## Goals

Here is a non-exhaustive list of target goals for this work:

### **Disable logic in-tact**

The existing "disable" logic should continue to work as it does now. Further, **disable logic should take priority over conditional logic** if a setting has both to evaluate. This serves two purposes:

1. This keeps the change backwards compatible, so we don't have to update every other option alongside this change. As needed, each setting can be updated to use the conditional logic. The only change to other settings may be to remove the "disable" logic from another setting in the scenario where you no longer want that disable logic to occur.
2. The current assumption the app makes is that a setting should always be disabled if there is any logic that could enable it but isn't explicitly doing so. This is probably a reasonably safe assumption to make - if one setting is set such that it is wildly incompatible with or without another setting, that other setting should probably be disabled. This assumption could certainly be challenged and updated in the future, but this change is (currently) focused on as little disturbance as possible.

### Enable logic: 

As the top of this PR described, we want to be able to "enable" a setting based on the state of other settings (one or multiple). This allows for something simple like:
- SettingA should be shown if SettingB is set to "true"

This also allows more complex logic like:
- SettingA should be shown if:
- - (SettingB is set to "true" OR SettingC is set to "peanuts") AND (SettingD is set to "false")

Still in the early phases, so this could be adapted much further. Obvious ideas are:
- Add the ability for this logic to both "enable" AND "disable" a setting (currently only supports enabling, and if no conditions are met it will be hidden).
- After adding the above note, use this to replace the current "disable" logic as it stands today.

## Feature Breakdown

- A setting's JSON can now be updated with a new dictionary object named "conditional_visibility".
- - Each entry in "conditional_visibility" is a "Condition". If a Condition is met for the setting, then the setting will be enabled and shown in the GUI. The key for each Condition object is a string for human-readable purposes only (for now).
- - - A Condition is a list of "Partial Conditions", an object of key/value pairs representing a different setting's name and a value for that setting. If at least one of the given settings is set to the respective value, the Partial Condition is considered to be met (OR logic).
- - For a Condition to be met, all child Partial Conditions must also be met (AND logic).
- A setting can have multiple Conditions. If any of them are met, then the setting will be enabled.
- If a setting has logic that can disable another setting, that disabling logic will take priority over any conditional logic the secondary setting may have.

### CODE REVIEW

File: GUI/src/app/pages/generator/generator.component.ts [hunk 1]
```diff
@@ -1264,6 +1264,27 @@ export class GeneratorComponent implements OnInit {
...
+            if (conditionsMetToEnable != this.global.generator_settingsVisibilityMap[dependentSetting.name]) {
+              console.log("Setting conditionally changing visibility: " + dependentSetting.name + " => " + conditionsMetToEnable)
+              this.global.generator_settingsVisibilityMap[dependentSetting.name] = conditionsMetToEnable;
+              triggeredChange = true;
+            }
+          }
+        });
+      }
+    
+      // This list of settings could be disabled entirely (Legacy)
```
Comment (LinkofOrigin) [inline GUI/src/app/pages/generator/generator.component.ts:1287]: Don't mind this next set of comments, they could be removed. Just there for my own sanity :)

File: GUI/src/app/pages/generator/generator.component.ts [hunk 2]
```diff
@@ -1272,19 +1293,59 @@ export class GeneratorComponent implements OnInit {
...
 
     return triggeredChange;
   }
 
+  private conditionsMetToEnable(settingConditions: any) {
+    let shouldEnable = false;
+    // There may be multiple combinations of conditions that may enable this setting.
+    // We'll check each one, and if one of them passes we'll enable the setting
+    for (let conditionName in settingConditions) {
+      console.log(conditionName + " => Evaluating condition...");
```
Comment (LinkofOrigin) [inline GUI/src/app/pages/generator/generator.component.ts:1319]: Fair number of `console.log` statements here. Just testing for now, I expect to remove these before any official merge.

File: SettingTypes.py [hunk 3]
```diff
@@ -10,7 +10,7 @@
 class SettingInfo:
     def __init__(self, setting_type: type, gui_text: Optional[str], gui_type: Optional[str], shared: bool,
                  choices: Optional[dict | list] = None, default: Any = None, disabled_default: Any = None,
-                 disable: Optional[dict] = None, gui_tooltip: Optional[str] = None, gui_params: Optional[dict] = None,
+                 disable: Optional[dict] = None, conditional_visibility: Optional[dict] = None, gui_tooltip: Optional[str] = None, gui_params: Optional[dict] = None,
```
Comment (LinkofOrigin) [inline SettingTypes.py:13]: New property for the class, just passed up the chain for the one class I bothered to update thus far. Remaining objects will need to be updated as well.

File: SettingsList.py [hunk 4]
```diff
@@ -716,8 +716,8 @@ class SettingInfos:
             'randomize_key': 'randomize_settings',
         },
         disable        = {
-            True:  {'settings': ['shuffle_ganon_bosskey', 'ganon_bosskey_stones', 'ganon_bosskey_medallions', 'ganon_bosskey_rewards', 'ganon_bosskey_tokens', 'ganon_bosskey_hearts']},
-            False: {'settings': ['triforce_count_per_world', 'triforce_goal_per_world']},
+            True:  {'settings': ['ganon_bosskey_stones', 'ganon_bosskey_medallions', 'ganon_bosskey_rewards', 'ganon_bosskey_tokens', 'ganon_bosskey_hearts']},
+            # False: {'settings': ['triforce_count_per_world', 'triforce_goal_per_world']},
```
Comment (LinkofOrigin) [inline SettingsList.py:720]: Now TriforceHunt no longer disables the ShuffleGanonBosskey setting when set to true (should it?). It also doesn't disable the Triforce*PerWorld settings when set to false, since now those rely on the "conditional_visibility" logic.

(PS. Yes, that commented line should be removed before merging. Too lazy to do so thus far since I have been swapping states back and forth a lot.)

File: SettingsList.py [hunk 5]
```diff
@@ -742,6 +742,18 @@ class SettingInfos:
...
+             # Single condition, name is for reading purposes only
+            "TCPW_ganon_boss_key_on_triforce_pieces_AND_triforce_hunt_disabled": [ # FIXME: EXAMPLE ONLY - REMOVE/UPDATE
+                # First partial condition that must be met for the whole condition to be true
+                {"shuffle_ganon_bosskey": "triforces"}, # List of setting/value pairs, of which at least ONE of which must be true (OR logic). key = setting name, value = setting value
+                # Second partial condition that must also be met for the whole condition to be true (AND logic)
+                {"triforce_hunt": False},
+            ],
+            "TCPW_triforce_hunt_enabled": [
+                {"triforce_hunt": True},
+            ],
```
Comment (LinkofOrigin) [inline SettingsList.py:755]: JUST FOR TESTING. Using this for demonstration and testing purposes, not actually suggesting this be part of the change. Will clean up before considering to merge.

Let's break it down:

> TCPW_ganon_boss_key_on_triforce_pieces_AND_triforce_hunt_disabled

Just a string name. Holds no functional purpose for now, but nice for debugging. I expect to change the structure of this at some point soon anyway, so not attached to this.

> {"shuffle_ganon_bosskey": "triforces"}

First entry in the condition's list - what I'll call a "partial condition". This partial condition requires the `shuffle_ganon_bosskey` setting to be set to the `triforces` value. If that is the case, this partial condition is met and considered "passing".

> {"triforce_hunt": False}

Second entry for this condition, separately evaluated as passing or not.

> "TCPW_ganon_boss_key_on_triforce_pieces_AND_triforce_hunt_disabled": [ 
>   {shuffle_ganon_bosskey": "triforces"},
>   {"triforce_hunt": False},
> ],

A full condition. Each of the partial conditions must pass for this whole condition to also pass. So `shuffle_ganon_bosskey` must be set to `triforces` AND `triforce_hunt` must be set to `false`. If that all passes, this setting will be displayed.

> "TCPW_ganon_boss_key_on_triforce_pieces_AND_triforce_hunt_disabled": [...],
> "TCPW_triforce_hunt_enabled": [...],

Two separate conditions for this "conditional_visibility" object. If at least one of them passes, this setting will be enabled!

File: SettingsList.py [hunk 6]
```diff
@@ -1069,6 +1093,7 @@ class SettingInfos:
             'dungeons':        "Dungeon Rewards",
             'tokens':          "Tokens",
             'hearts':          "Hearts",
+            'triforces':       "Triforce Pieces",
```
Comment (LinkofOrigin) [inline SettingsList.py:1096]: TESTING ONLY. Just a new fake setting to demonstrate/test with. (and maybe a pipe dream :3)

File: SettingsList.py [hunk 7]
```diff
@@ -1118,6 +1145,7 @@ class SettingInfos:
             '!dungeons':    {'settings': ['ganon_bosskey_rewards']},
             '!tokens':      {'settings': ['ganon_bosskey_tokens']},
             '!hearts':      {'settings': ['ganon_bosskey_hearts']},
+            # 'triforces':    {'settings': ['ganon_bosskey_stones', 'ganon_bosskey_medallions','ganon_bosskey_rewards','ganon_bosskey_tokens','ganon_bosskey_hearts']},
```
Comment (LinkofOrigin) [inline SettingsList.py:1148]: Same as the other above, still testing and been too lazy to delete/revert. Will be cleaned up later.

File: SettingsList.py [hunk 8]
```diff
@@ -5575,9 +5603,9 @@ class UnmappedSettingError(Exception):
 
     if info.disable is not None:
         for option, disabling in info.disable.items():
-            negative = False
+            negative = False # If this option is enabled, the "disabling" settings will be disabled
             if isinstance(option, str) and option[0] == '!':
-                negative = True
+                negative = True # If this option is NOT enabled, the "disabling" settings will be disabled
```
Comment (LinkofOrigin) [inline SettingsList.py:5608]: Sanity-saving comments.

File: SettingsToJson.py [hunk 9]
```diff
@@ -53,6 +54,11 @@ def add_disable_option_to_json(disable_option: dict[str, Any], option_json: dict
             option_json['controls_visibility_tab'] += ',' + ','.join(disable_option['tabs'])
 
 
+def mark_conditional_visibility_option_for_setting(setting_name: str, conditional_settings: dict[str, Any]) -> None:
+    if setting_name not in conditional_visibility_dependencies:
+        conditional_visibility_dependencies[setting_name] = conditional_settings
```
Comment (LinkofOrigin) [inline SettingsToJson.py:59]: We need to build the full settings JSON before we can inject any conditional data into it. We would either need to inject early and be sure to carefully update it later, or do this where we deal with it at the end. This felt like the least disruptive option for now.

File: SettingsToJson.py [hunk 10]
```diff
@@ -82,6 +88,11 @@ def get_setting_json(setting: str, web_version: bool, as_array: bool = False) ->
     setting_disable = {}
     if setting_info.disable is not None:
         setting_disable = copy.deepcopy(setting_info.disable)
+    
+    # If this setting can be conditionally enabled, we will need to revisit it once the full JSON has been built
+    if setting_info.conditional_visibility is not None:
+        mark_conditional_visibility_option_for_setting(setting_info.name, setting_info.conditional_visibility)
+        setting_json['conditional_visibility'] = setting_info.conditional_visibility
```
Comment (LinkofOrigin) [inline SettingsToJson.py:95]: Save this to be processed for later and also save it to this setting's JSON data. When the aforementioned "later" comes, we'll update the other settings to denote that they can alter this setting.

File: SettingsToJson.py [hunk 11]
```diff
@@ -260,6 +271,9 @@ def create_settings_list_json(path: str, web_version: bool = False) -> None:
             output_json['cosmeticsObj'][tab['name']] = tab_json_object
             output_json['cosmeticsArray'].append(tab_json_array)
 
+    # Resolve conditional visibility settings now that the full json is available
+    resolve_conditional_visibility_dependencies(output_json)
```
Comment (LinkofOrigin) [inline SettingsToJson.py:275]: The previously aforementioned "later" has arrived.

File: SettingsToJson.py [hunk 12]
```diff
@@ -273,6 +287,51 @@ def create_settings_list_json(path: str, web_version: bool = False) -> None:
         json.dump(output_json, f)
 
 
+def resolve_conditional_visibility_dependencies(output_json: dict) -> None:
```
Comment (LinkofOrigin) [inline SettingsToJson.py:290]: Really ugly nesting logic here. I'm directly walking through conditional logic we saved for later AND the actual json object we built up for all the settings. This is so we can locate the settings that could enable other settings per any conditional logic.

With this, we update those settings' options with a new `conditionally_controls_setting` entry with a list of settings that it may alter when set. We do this to distinguish it from `controls_visibility_setting` since that has a very baked-in set of logic to disable a setting. The new logic could also enable it, so we need to handle that too.
